### ISSUE TITLE
no-extend-native false positive 

### ISSUE THREAD

**Issue by wesleytodd** [eslint/eslint#8461]
Applying `standard@10.0.1` to the npm package [`migrate`](https://github.com/tj/node-migrate) and it has a constructor called `Set`.  Obviously it might not be the best name since it is a native constructor now, but eslint should probably recognize that it is not actually modifying the native.

```
Projects/node-migrate/lib/set.js:52:1: Set prototype is read only, properties should not be added.
```

```js
function Set (path) {}
Set.prototype.addMigration = function (title, up, down) {}

**Comment by platinumazure** [comment]
I've only had a chance to skim the rule source, but I think I agree this is a bug. The rule is looking at built-in global names without using scope analysis to determine if a global scope reference is even being modified (which it clearly isn't in the example case).

**Comment by platinumazure** [comment]
I'll try to repro when I get on my laptop, and I'll try to get a fix in sometime this weekend.

**Comment by not-an-aardvark** [comment]
If `Set` is declared at the top level, wouldn't this result in reassigning the global?

**Comment by wesleytodd** [comment]
@not-an-aardvark Yes, but since this is in a module, it is scoped to the module function scope.  In my fork I changed the name to `MigrationSet` because I agree that it is better to just avoid the issue altogether.  If I had left it as `Set` it would mean you cannot access the global `Set` inside that module.

**Comment by platinumazure** [comment]
Sure, but that's not what this rule is for. This rule is about extending
the prototype of built-in objects. Also, in a Node module, no, it wouldn't
reassign the global.

**Comment by not-an-aardvark** [comment]
Sorry, I got this rule confused with `no-global-assign`

**Comment by platinumazure** [comment]
This example also results in a lint error:

```js
/* eslint-env es6 */
{
    let Set = function() {};
    Set.prototype.add = function() {};
}
```

This very clearly doesn't extend the global Set object, because I have shadowed it with a block-scope variable. So I think this is definitely a bug.

**Comment by platinumazure** [comment]
Update: I'm about halfway through rewriting the rule to use eslint-scope to do scope analysis, but I got quite busy over the weekend. I'm hoping to have a fix for this in time for this week's planned alpha release on Friday.

**Comment by wesleytodd** [comment]
Wow, awesomely quick turn around on this.  Thanks @platinumazure!!

**Comment by platinumazure** [comment]
@wesleytodd Save your thanks until I can finish the job :smile:

**Comment by platinumazure** [comment]
Update: No meaningful updates, been busy the last few days. Unfortunately this won't make tomorrow's release, but it'll get in before 4.0.0 is out.

**Comment by platinumazure** [comment]
Finished the basic changes, just need to debug some failing test cases. Hoping to have a PR up this week.

**Comment by platinumazure** [comment]
@wesleytodd PR is up: #8528. Sorry for taking so long with this!

**Comment by wesleytodd** [comment]
I am not familiar enough with the codebase to meaningfully contribute to the conversation, but looks good to me!  Thanks @platinumazure for the quick work on this bug, it was not "so long" at all!

I will close this, but feel free to re-open it if you all need it for tracking purposes.  Thanks!

**Comment by not-an-aardvark** [comment]
Thanks @wesleytodd. I'll reopen this for tracking purposes just in case the PR ends up getting delayed/blocked for some reason (which I don't think will happen, but it's probably worth keeping the issue open anyway).
